### PR TITLE
In the checkoutBranch method, move stash logic into new method.

### PR DESCRIPTION
**Pull request by probablycorey**
## Overview

Closes https://github.com/desktop/desktop/issues/7609

## Description

This is a straight forward extraction of some code in the _checkoutBranch method . 

- I ended up with a method name that is pretty long `_stashBasedOnUncommittedChangesStrategy` so if you have a better name I'm all ears 🌽! 
- I removed the `enableStashing` because the feature is globally enabled!

### CODE REVIEW

File: app/src/lib/stores/app-store.ts [hunk 1]
```diff
@@ -3037,6 +3000,50 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
   }
 
+  private async _stashBasedOnUncommittedChangesStrategy(
+    repository: Repository,
+    branch: Branch,
+    uncommittedChangesStrategy: UncommittedChangesStrategy = askToStash
+  ): Promise<IStashEntry | null> {
+    const {
+      changesState,
+      branchesState: { tip },
```
Comment (kuychaco) [inline app/src/lib/stores/app-store.ts]: ✨oooh nested destructuring... I forget about this

File: app/src/lib/stores/app-store.ts [hunk 2]
```diff
@@ -2907,59 +2907,22 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+    if (hasChanges && uncommittedChangesStrategy.kind === askToStash.kind) {
+      this._showPopup({
+        type: PopupType.StashAndSwitchBranch,
+        branchToCheckout: foundBranch,
+        repository,
+      })
+      return repository
     }
 
+    let stashToPop: IStashEntry | null = await this._stashBasedOnUncommittedChangesStrategy(
```
Comment (kuychaco) [inline app/src/lib/stores/app-store.ts]: With this extraction, it would be awesome to make this a `const` and minimize mutation if not completely necessary. Which I think we should be able to do with:

```
diff --git a/app/src/lib/stores/app-store.ts b/app/src/lib/stores/app-store.ts
index 0d7d02af4..152dbc75a 100644
--- a/app/src/lib/stores/app-store.ts
+++ b/app/src/lib/stores/app-store.ts
@@ -2958,9 +2958,10 @@ export class AppStore extends TypedBaseStore<IAppState> {
       // call this method again once the working directory has been cleared.
       this.statsStore.recordChangesTakenToNewBranch()
 
-      stashToPop = stashToPop || uncommittedChangesStrategy.transientStashEntry
-      if (stashToPop !== null) {
-        const stashSha = stashToPop.stashSha
+      const stashEntry =
+        stashToPop || uncommittedChangesStrategy.transientStashEntry
+      if (stashEntry !== null) {
+        const stashSha = stashEntry.stashSha
         await gitStore.performFailableOperation(() => {
           return popStashEntry(repository, stashSha)
         })
```

This'll make it easier to scan the code and know with certainty what `stashToPop` is (since it's value is assigned in only one place) and we can ensure that we don't add logic in the future that changes the value in unexpected ways.

Comment (outofambit) [reply app/src/lib/stores/app-store.ts]: yeah, i agree this would be a nice improvement too! also agree its non-blocking.

fwiw this line was only moved down from earlier in the function by this PR (see https://github.com/desktop/desktop/pull/8087/files#diff-d6b76b602a5c48740d6d10deda6f71ebL2910).

File: app/src/lib/stores/app-store.ts (hunk 1 above)
Comment (probablycorey) [reply app/src/lib/stores/app-store.ts]: I didn't know if this was too weird. It's kind of hard to follow and spitting it into two lines might be more readable.

**Comment by probablycorey** [comment]
> My first small request is that we let the enableStashing check live on as a guard to not show the popup when there are changes...

I'm less worried about the organic decay of feature-flag methods, but I'm deferring to your knowledge on this one!

> I think we should look into having the stashBasedOnUncommittedChangesStrategy return the transientStashEntry if there is one

This was annoying me too. I switched it around to do what you suggested. `transientStashEntry` was confusing to me, so hiding it behind a cleaner stash interface makes sense to me.

> Lastly, while we don't require documentation for private methods I feel like this one warrants it,

I couldn't come up with a great new method name, so I updated the name to `stashToPopAfterBranchCheckout` and added a comment.
